builtin/logical/database/rollback.go

```diff
@@ -0,0 +1,115 @@
+	data interface{}) error {
+
+	if kind != rotateRootWALKey {
+		return errors.New("unknown type to rollback")
+	}
+
+	// Decode the WAL data
+	var entry rotateRootCredentialsWAL
+	if err := mapstructure.Decode(data, &entry); err != nil {
+		return err
+	}
+
+	// Get the current database configuration from storage
+	config, err := b.DatabaseConfig(ctx, req.Storage, entry.ConnectionName)
+	if err != nil {
+		return err
+	}
+
+	// The password in storage doesn't match the new password
+	// in the WAL entry. This means there was a partial failure
+	// to update either the database or storage.
+	if config.ConnectionDetails["password"] != entry.NewPassword {
+		// Clear any cached connection to inform the rollback decision
+		err := b.ClearConnection(entry.ConnectionName)
+		if err != nil {
+			return err
+		}
+
+		// Attempt to get a connection with the current configuration.
+		// If successful, the WAL entry can be deleted. This means
+		// the root credentials are the same according to the database
+		// and storage.
+		_, err = b.GetConnection(ctx, req.Storage, entry.ConnectionName)
+		if err == nil {
+			return nil
+		}
+
+		return b.rollbackDatabaseCredentials(ctx, config, entry)
+	}
+
+	// The password in storage matches the new password in
+	// the WAL entry, so there is nothing to roll back. This
+	// means the new password was successfully updated in the
+	// database and storage, but the WAL wasn't deleted.
+	return nil
+}
+
+// rollbackDatabaseCredentials rolls back root database credentials for
+// the connection associated with the passed WAL entry. It will creates
+// a connection to the database using the WAL entry new password in
+// order to alter the password to be the WAL entry old password.
+func (b *databaseBackend) rollbackDatabaseCredentials(ctx context.Context, config *DatabaseConfig,
+	entry rotateRootCredentialsWAL) error {
+
+	// Attempt to get a connection with the WAL entry new password.
+	config.ConnectionDetails["password"] = entry.NewPassword
+	dbc, err := b.GetConnectionWithConfig(ctx, entry.ConnectionName, config)
+	if err != nil {
+		return err
+	}
+
+	// Ensure the connection used to roll back the database password is not cached
+	defer func() {
+		if err := b.ClearConnection(entry.ConnectionName); err != nil {
+			b.Logger().Error("error closing database plugin connection", "err", err)
+		}
+	}()
+
+	// Roll back the database password to the WAL entry old password
+	statements := dbplugin.Statements{Rotation: config.RootCredentialsRotateStatements}
+	userConfig := dbplugin.StaticUserConfig{
+		Username: entry.UserName,
+		Password: entry.OldPassword,
+	}
+	_, _, err = dbc.SetCredentials(ctx, statements, userConfig)
+
+	// If the database plugin doesn't implement SetCredentials,
+	// the root credentials can't be rolled back. This means
+	// the root credential rotation happened via the plugin
+	// RotateRootCredentials RPC.
+	if err != nil && status.Code(err) == codes.Unimplemented {
```

Comment:
Same question here around wondering if we can slightly rewrite this to make the err handling more at-a-glance.

Reply:
Will take your suggestion from above and see if I can make this easier to read.

Reply:
Done in 03c73d9. Let me know your thoughts.

Reply:
Lovely, thanks!